examples/vqe.lisp

```diff
@@ -0,0 +1,236 @@
+                                    -0.36914431524376695d0
+                                     0.17992650976405983d0
+                                     0.042217556922433896d0
+                                    -0.042217556922433896d0
+                                    -0.042217556922433896d0
+                                     0.042217556922433896d0
+                                     0.13459240346368873d0
+                                     0.17680996038612262d0
+                                     0.17680996038612262d0
+                                     0.13459240346368873d0
+                                     0.18620984259247159d0)
+  "Coefficients of the Pauli operators present in the Hamiltonian.")
+
+(defparameter *pauli-terms* (mapcar #'quil:parse-quil '("I 0"
+                                                        "Z 0"
+                                                        "Z 1"
+                                                        "Z 2"
+                                                        "Z 3"
+                                                        "Z 0; Z 1"
+                                                        "Y 0; X 1; X 2; Y 3"
+                                                        "X 0; X 1; Y 2; Y 3"
+                                                        "Y 0; Y 1; X 2; X 3"
+                                                        "X 0; Y 1; Y 2; X 3"
+                                                        "Z 0; Z 2"
+                                                        "Z 0; Z 3"
+                                                        "Z 1; Z 2"
+                                                        "Z 1; Z 3"
+                                                        "Z 2; Z 3"))
+  "Pauli operators in the Hamiltonian.")
+
+(defparameter *initial-thetas* #(0.0d0 -0.036014483d0)
+  "Initial guess for the classical optimization method.")
+
+(defun run-vqe (&key (number-of-qubits *number-of-qubits*)
+                  (initial-thetas *initial-thetas*)
+                  (reference-state *reference-state*)
+                  (ansatz *ansatz*)
+                  (pauli-coefficients *pauli-coefficients*)
+                  (pauli-terms *pauli-terms*))
+  "Run the variational quantum eigensolver algorithm."
+  (flet ((objective-function (thetas)
+           "Compute the energy based on the values of THETAS."
+           (loop :with state-prep := (quil:parse-quil
+                                      (format nil "DECLARE theta REAL[2]~%~
+                                                   MOVE theta[0] ~F~%~
+                                                   MOVE theta[1] ~F~%~A~A"
+                                              (elt thetas 0) (elt thetas 1)
+                                              reference-state ansatz))
+                 :with expectations := (qvm-app::perform-expectation 'qvm-app::pure-state state-prep pauli-terms number-of-qubits)
```

Comment:
I'm not saying it needs to happen in this PR, but maybe the fact that `qvm-examples` needs this function is an indication that something like `perform-expectation` could be useful to have in the `qvm` package, rather than needing to depend on `qvm-app` here. I guess that would also break the dependency cycle.

Reply:
I agree with your comment above with regard to copying and pasting and I have posted #185 for the time being.

With respect to `qvm-app`, the question is less clear: if `qvm-app` is only meant to be an executable entry point to the QVM, then I concur that the code in `app/src/expectation.lisp` belongs in the `qvm` package instead. However, if the scope of `qvm-app` also encompasses high-level functionality built on top of the QVM (like multishot measures, etc.), then I it should stay where it currently is.

Reply:
Thanks!

My gut feeling is that `qvm-app` *should* be just the executable entry-point + RPC API etc. But you're right that it currently also has some higher-level stuff that might be useful elsewhere. The fact that it includes command line processing + main + hunchentoot makes it an awkward general-purpose library though, in my opinion. Maybe that means some of the stuff in `qvm-app` could be pushed down into `qvm`, or maybe there should be a new package. Maybe this really falls into my court with the `qvm-app-ng` redesign :-).

Reply:
Let's get expectation value computation into the QVM proper. It seems reasonable. @jmbr could you file an issue for that?

Reply:
I just posted this issue: #192 .